lib/rspec/core/set.rb

```diff
@@ -0,0 +1,41 @@
+module RSpec
+  module Core
+    # @private
+    #
+    # We use this to replace `::Set` so we can have the advantage of
+    # constant time key lookups for unique arrays but without the
+    # potential to pollute a developers environment with an extra
+    # peice of the stdlib. This helps to prevent false positive
+    # builds.
+    #
+    class Set
+      include Enumerable
+
+      def initialize(array=[])
+        @values = {}
+        merge(array)
+      end
+
+      def <<(key)
+        @values[key] = true
+        self
+      end
+
+      def each(&block)
+        @values.keys.each(&block)
```

Comment:
> I have just now, #keys.each is faster but not by a lot.

Weird.  I wonder why `each_key` even exists then...given that all it does is yield each key I'd expect it to be optimized for that case and be faster than the hop of `keys.each`.  Can you commit your benchmark for future reference?